File: scraper/python/GrabzIt/GrabzItScrapeClient.py
#!/usr/bin/python

import os

try:
	from urllib.parse import urlencode	
	from urllib.request import urlopen
	from urllib.parse import urlparse
	import http.client as httpClient
except ImportError:
	from urllib import urlopen
	from urlparse import urlparse	
	from urllib import urlencode	
	import httplib as httpClient

import hashlib	
import mimetypes
import io
from xml.dom import minidom
from time import sleep
from GrabzIt import GrabzItScrape
from GrabzIt import GrabzItScrapeHistory
from GrabzIt import GrabzItScrapeException
from GrabzIt import GrabzItTarget
from GrabzIt import GrabzItVariable

class GrabzItScrapeClient:

		WebServicesBaseURL = "/services/scraper/"
		TrueString = "True"

		def __init__(self, applicationKey, applicationSecret):
				self.applicationKey = applicationKey
				self.applicationSecret = applicationSecret	 
				self.proxy = None				 
		
		#
		#This method enables a local proxy server to be used for all requests
		#
		#proxyUrl - the URL, which can include a port if required, of the proxy. Providing a null will remove any previously set proxy
		#
		def SetLocalProxy(self, proxyUrl):
			self.proxy = proxyUrl		 
		
		#
		#Get all of the users scrapes
		#
		#id - If specified, just returns that scrape that matches the id.
		#
		#This function returns a array of Scrape objects
		#
		def GetScrapes(self, id = ""):
				sig =  self.CreateSignature(str(self.applicationSecret)+"|"+str(id))
				qs = {"key":self.applicationKey, "identifier":id}
				encoded_qs = urlencode(qs)
				
				encoded_qs += "&sig="+sig

				dom = minidom.parseString(self.HTTPGet(self.WebServicesBaseURL + "getscrapes?" + encoded_qs))

				self.CheckForException(dom)
						
				results = []

				nodes = dom.getElementsByTagName("Scrape")
				
				for node in nodes:
						identifier = node.getElementsByTagName('Identifier')
						name = node.getElementsByTagName('Name')
						status = node.getElementsByTagName('Status')
						nextRun = node.getElementsByTagName('NextRun')
						
						scrapeResults = []
						resultNodes = node.getElementsByTagName('Result')
						
						for resultNode in resultNodes:
							resultIdentifier = resultNode.getElementsByTagName('Identifier')
							resultFinished = resultNode.getElementsByTagName('Finished')
							scrapeResults.append(GrabzItScrapeHistory.GrabzItScrapeHistory(self.GetFirstValue(resultIdentifier), self.GetFirstValue(resultFinished)))
							
						results.append(GrabzItScrape.GrabzItScrape(self.GetFirstValue(identifier), self.GetFirstValue(name), self.GetFirstValue(status), self.GetFirstValue(nextRun), scrapeResults))																 

				return results
				
		#
		#Sets the status of a scrape. 
		#
		#id - The id of the scrape to set.
		#status - The scrape status to set the scrape to. Options include Start, Stop, Enable and Disable
		#
		#This function returns true if the scrape was successfully set.
		#
		def SetScrapeStatus(self, id, status):
				sig = self.CreateSignature(str(self.applicationSecret)+"|"+str(id)+"|"+str(status))

				qs = {"key":self.applicationKey, "id":id, "status":status}

				encoded_qs = urlencode(qs)
				
				encoded_qs += "&sig="+sig;

				return self.IsSuccessful(self.HTTPGet(self.WebServicesBaseURL + "setscrapestatus?" + encoded_qs))								 
				
		#
		#Set a property of a scrape
		#
		#id - The id of the scrape to set.
		#property - The property object that contains the required changes
		#
		#This function returns true if successful
		#
		def SetScrapeProperty(self, id, property):
			if (id == None or property == None):
				return False
				
			sig = self.CreateSignature(str(self.applicationSecret)+"|"+str(id)+"|"+str(property.GetTypeName()))
			
			fields = {}
			fields['key'] = self.applicationKey
			fields['id'] = str(id)
			fields['payload'] = property.ToXML()
			fields['type'] = property.GetTypeName()
			fields['sig'] = sig
			
			return self.IsSuccessful(self.HTTPPost("/services/scraper/setscrapeproperty", fields))
		#
		#Re-sends the scrape result with the matching scrape id and result id using the export parameters stored in the scrape. 
		#
		#id - The id of the scrape that contains the result to re-send.
		#resultId - The id of the result to re-send.
		#
		#This function returns true if the result was successfully requested to be re-sent.
		#
		def SendResult(self, id, resultId):
				sig =  self.CreateSignature(str(self.applicationSecret)+"|"+str(id)+"|"+str(resultId))		  

				qs = {"key":self.applicationKey, "id":id, "spiderId":resultId}

				encoded_qs = urlencode(qs)
				
				encoded_qs += "&sig="+sig;

				return self.IsSuccessful(self.HTTPGet(self.WebServicesBaseURL + "sendscrape?" + encoded_qs))				 
				
		
		def IsSuccessful(self, result):
				return self.GetResultObject(result, "Result") == self.TrueString
		
		def GetResultObject(self, result, resultTagName):
				dom = minidom.parseString(result)
				
				result = ""
								
				nodes = dom.getElementsByTagName(resultTagName)
				
				self.CheckForException(dom)
				
				for node in nodes:
						if node.firstChild == None:
								break
						result = node.firstChild.nodeValue
						break					
										
				return result
 
		def CheckForException(self, dom):
				if dom == None:
						return
				
				message = ""
				code = ""				
				
				messageNodes = dom.getElementsByTagName("Message")
				codeNodes = dom.getElementsByTagName("Code")
				
				for messageNode in messageNodes:
						if messageNode.firstChild == None:
								break
						message = messageNode.firstChild.nodeValue
						break

				for codeNode in codeNodes:
						if codeNode.firstChild == None:
								break
						code = codeNode.firstChild.nodeValue
						break
				
				if len(message) > 0:
						raise GrabzItScrapeException.GrabzItScrapeException(message, code)			
	   
				
		def GetFirstValue(self, node):
				if node.length > 0 and node[0].firstChild:
						return node[0].firstChild.nodeValue
				return ""
		
		def HTTPGet(self, url):
				connection = self._getConnection('GET', url)
				connection.endheaders()
				response = connection.getresponse()
				self.CheckResponseHeader(response.status)
				return response.read()	
				
		def HTTPPost(self, selector, fields):
			body = urlencode(fields)
					
			h = self._getConnection('POST', selector)
			h.putheader('content-type', "application/x-www-form-urlencoded")
			h.putheader('content-length', str(len(body)))
			h.endheaders()
			
			try:
					h.send(body)
			except TypeError:
					# python 3 needs it to be encoded
					h.send(body.encode())
					
			response = h.getresponse()
			self.CheckResponseHeader(response.status);
			return response.read()				  

		def _getConnection(self, method, action):
			hostname = "grabz.it"
			port = 80
			username = None
			password = None
			authHeader = None
						
			if self.proxy:
				url = urlparse(self.proxy)
				hostname = url.hostname
				port = url.port					 
				username = unquote(url.username)
				password = unquote(url.password)
					
			h = httpClient.HTTPConnection(hostname, port)
			
			if self.proxy:
				if username and password:
					auth = '%s:%s' % (username, password)
					
					try:
						authHeader = base64.b64encode(auth)
					except TypeError:
						authHeader = base64.b64encode(auth.encode()).decode("ISO-8859-1")
						
				h.putheader('proxy-authorization', 'Basic ' + authHeader)
				
				port = 80			  
				h.set_tunnel('grabz.it', port)
			
			h.putrequest(method, action)

			return h
			
		def CheckResponseHeader(self, httpCode):
				if httpCode == 403:
						raise GrabzItScrapeException.GrabzItScrapeException('Potential DDOS Attack Detected. Please wait for your service to resume shortly. Also please slow the rate of requests you are sending to GrabzIt to ensure this does not happen in the future.', GrabzItException.GrabzItException.NETWORK_DDOS_ATTACK)
				elif httpCode >= 400:
						raise GrabzItScrapeException.GrabzItScrapeException('A network error occured when connecting to the GrabzIt servers.', GrabzItException.GrabzItException.NETWORK_GENERAL_ERROR)

		def CreateSignature(self, value):
				md5 = hashlib.md5()
				md5.update(value.encode('ascii', 'replace'))
				return md5.hexdigest()

File: scraper/php/lib/GrabzItScrapeClient.php
<?php
namespace GrabzIt\Scraper;

spl_autoload_register(function ($class_name) {
	$file_name = str_replace("GrabzIt\\Scraper\\", '', $class_name) . '.php';
	if (strpos($file_name, "GrabzIt") === 0) {
		include($file_name);
	}
});

class GrabzItScrapeClient
{
	const WebServicesBaseURL = "http://api.grabz.it/services/scraper/";
	const TrueString = "True";

	private $applicationKey;
	private $applicationSecret;
	private $connectionTimeout = 600;
	private $proxy = null;

	public function __construct($applicationKey, $applicationSecret)
	{
		$this->applicationKey = $applicationKey;
		$this->applicationSecret = $applicationSecret;
	}
	
	/*
	This method enables a local proxy server to be used for all requests
	
	proxyUrl - the URL, which can include a port if required, of the proxy. Providing a null will remove any previously set proxy
	*/
	public function SetLocalProxy($proxyUrl)
	{
		if ($proxyUrl == null || empty($proxy))
		{
			$this->proxy = null;
			return;
		}
		
		$url_parts = parse_url($proxyUrl);
		
		$this->proxy = new stdClass();
		$this->proxy->host = $url_parts['host'];
		$this->proxy->port = $url_parts['port'];
		$this->proxy->username = null;
		$this->proxy->password = null;
		
		if (isset($url_parts['user']))
		{
			$this->proxy->username = urldecode($url_parts['user']);
		}
		if (isset($url_parts['pass']))
		{
			$this->proxy->password = urldecode($url_parts['pass']);
		}
	}	

	public function SetTimeout($timeout)
	{
		$this->connectionTimeout = $timeout;
	}

	public function SetApplicationKey($applicationKey)
	{
		$this->applicationKey = $applicationKey;
	}

	public function GetApplicationKey()
	{
		return $this->applicationKey;
	}

	public function SetApplicationSecret($applicationSecret)
	{
		$this->applicationSecret = $applicationSecret;
	}

	public function GetApplicationSecret()
	{
		return $this->applicationSecret;
	}

	/*
	Get all of the requested scrapes

	id - If specified, just returns that scrape that matches the id.

	This function returns a array of Scrape objects
	*/
	public function GetScrapes($id = "")
	{
		$sig =	$this->encode($this->applicationSecret."|".$id);

		$qs = "key=" .urlencode($this->applicationKey)."&identifier=".$id."&sig=".$sig;

		$obj = $this->getResultObject($this->Get(GrabzItScrapeClient::WebServicesBaseURL . "getscrapes.ashx?" . $qs));

		$result = array();

		if ($obj == null)
		{
			return $result;
		}
		
		foreach ($obj->Scrapes->Scrape as $scrape)
		{
			$grabzitScrape = new GrabzItScrape();
			$grabzitScrape->ID = (string)$scrape->Identifier;
			$grabzitScrape->Name = (string)$scrape->Name;
			$grabzitScrape->Status = (string)$scrape->Status;
			$grabzitScrape->NextRun = (string)$scrape->NextRun;
			foreach ($scrape->Results->Result as $scrapeResult)
			{
				$grabzItScrapeHistory = new GrabzItScrapeHistory();
				$grabzItScrapeHistory->ID = (string)$scrapeResult->Identifier;
				$grabzItScrapeHistory->Finished = (string)$scrapeResult->Finished;
				
				$grabzitScrape->Results[] = $grabzItScrapeHistory;
			}
			
			$result[] = $grabzitScrape;
		}

		return $result;
	}

	/*
	Sets the status of a scrape.

	id - The id of the scrape to set.
	status - The scrape status to set the scrape to. Options include Start, Stop, Enable and Disable

	This function returns true if the scrape was successfully set.
	*/
	public function SetScrapeStatus($id, $status)
	{
		if (empty($id))
		{
			return false;
		}

		$sig =	$this->encode($this->applicationSecret."|".$id."|".$status);

		$qs = "key=" .urlencode($this->applicationKey)."&id=".urlencode($id)."&status=".urlencode($status)."&sig=".$sig;

		return $this->isSuccessful($this->Get(GrabzItScrapeClient::WebServicesBaseURL . "setscrapestatus.ashx?" . $qs));
	}
	
	/*
	Set a property of a scrape

	id - The id of the scrape to set.
	property - The property object that contains the required changes

	This function returns true if successful
	*/
	public function SetScrapeProperty($id, $property)
	{
		if (empty($id) || $property == null)
		{
			return false;
		}
		
		$sig =	$this->encode($this->applicationSecret."|".$id."|".$property->GetTypeName());
		
		$qs = array();
		$qs["key"] = $this->applicationKey;
		$qs["id"] = $id;
		$qs["payload"] = $property->ToXML();
		$qs["type"] = $property->GetTypeName();
		$qs["sig"] = $sig;

		return $this->isSuccessful($this->Post(GrabzItScrapeClient::WebServicesBaseURL . "setscrapeproperty.ashx", $qs));
	}
	
	/*
	Re-sends the scrape result with the matching scrape id and result id using the export parameters stored in the scrape.

	id - The id of the scrape that contains the result to re-send.
	resultId - The id of the result to re-send.

	This function returns true if the result was successfully requested to be re-sent.
	*/	  
	public function SendResult($id, $resultId)
	{
		if (empty($id) || empty($resultId))
		{
			return false;
		}

		$sig =	$this->encode($this->applicationSecret."|".$id."|".$resultId);

		$qs = "key=" .urlencode($this->applicationKey)."&id=".urlencode($id)."&spiderId=".urlencode($resultId)."&sig=".$sig;

		return $this->isSuccessful($this->Get(GrabzItScrapeClient::WebServicesBaseURL . "sendscrape.ashx?" . $qs));
	}	 

	private function isSuccessful($result)
	{
		$obj = $this->getResultObject($result);
		return ((string)$obj->Result == GrabzItScrapeClient::TrueString);
	}

	private function getResultObject($result)
	{
		$obj = simplexml_load_string($result);

		if (!empty($obj->Message))
		{
			throw new GrabzItScrapeException($obj->Message, $obj->Code);
		}

		return $obj;
	}

	private function encode($text)
	{
		return md5(mb_convert_encoding($text, "ASCII", mb_detect_encoding($text)));
	}

	private function Get($url)
	{
		if (ini_get('allow_url_fopen'))
		{
			$opts = array('http' => array('timeout' => $this->connectionTimeout));
			$opts = $this->addProxyToStreamContext($opts);
			$context = stream_context_create($opts);
			$response = @file_get_contents($url, false, $context);

			if (isset($http_response_header))
			{
				$this->checkResponseHeader($http_response_header);
			}

			return $response;
		}

		if (function_exists('curl_version'))
		{
			$ch = curl_init();
			curl_setopt($ch,CURLOPT_URL,$url);
			curl_setopt($ch,CURLOPT_RETURNTRANSFER,1);
			curl_setopt($ch,CURLOPT_CONNECTTIMEOUT,$this->connectionTimeout);
			$this->addProxyToCurl($ch);
			
			$data = curl_exec($ch);
			$httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);

			$this->checkHttpCode($httpCode);

			curl_close($ch);

			return $data;
		}
		throw new GrabzItScrapeException("Unable to contact GrabzIt's servers. Please install the CURL extension or set allow_url_fopen to 1 in the php.ini file.", GrabzItException::GENERIC_ERROR);
	}

	private function Post($url, $parameters)
	{
		if (ini_get('allow_url_fopen'))
		{
			$options = array(
				'http' => array(
					'timeout' => $this->connectionTimeout,
					'header'  => "Content-type: application/x-www-form-urlencoded\r\n",
					'method'  => 'POST',
					'content' => http_build_query($parameters, '', '&')
				)
			);

			$options = $this->addProxyToStreamContext($options);
			$context  = stream_context_create($options);
			$response = @file_get_contents($url, false, $context);

			if (isset($http_response_header))
			{
				$this->checkResponseHeader($http_response_header);
			}
			
			if ($response === FALSE)
			{
				throw new GrabzItScrapeException("An unknown network error occurred.", GrabzItException::NETWORK_GENERAL_ERROR);
			}

			return $response;
		}
		
		if (function_exists('curl_version'))
		{
			$ch = curl_init();

			//set the url, number of POST vars, POST data
			curl_setopt($ch,CURLOPT_URL, $url);
			curl_setopt($ch,CURLOPT_POST, count($parameters));
			curl_setopt($ch,CURLOPT_CONNECTTIMEOUT,$this->connectionTimeout);
			curl_setopt($ch,CURLOPT_POSTFIELDS, http_build_query($parameters, '', '&'));
			$this->addProxyToCurl($ch);

			//execute post
			$data = curl_exec($ch);

			$httpCode = curl_getinfo($ch, CURLINFO_HTTP_CODE);

			$this->checkHttpCode($httpCode);

			//close connection
			curl_close($ch);

			return $data;
		}
		
		throw new GrabzItScrapeException("Unable to contact GrabzIt's servers. Please install the CURL extension or set allow_url_fopen to 1 in the php.ini file.", GrabzItScrapeException::GENERIC_ERROR);
	}
	
	private function addProxyToCurl($ch)
	{
		if ($this->proxy != null)
		{
			curl_setopt($ch, CURLOPT_PROXYPORT, $this->proxy->port);
			curl_setopt($ch, CURLOPT_PROXYTYPE, 'HTTP');
			curl_setopt($ch, CURLOPT_PROXY, $this->proxy->host);
			if ($this->proxy->username != null && $this->proxy->password != null)
			{
				curl_setopt($ch, CURLOPT_PROXYUSERPWD, base64_encode($this->proxy->username . ':' . $this->proxy->password));
			}
		}		
	}
	
	private function addProxyToStreamContext($options)
	{
		if ($this->proxy != null)
		{
			$options['http']['request_fulluri'] = true;
			$options['http']['proxy'] = $this->proxy->host . ':' . $this->proxy->port;
			if ($this->proxy->username != null && $this->proxy->password != null)
			{
				$headers = array();
				if (isset($options['http']['header']))
				{
					$headers = $options['http']['header'];
				}
				array_push($headers, 'Proxy-Authorization: Basic' . base64_encode($this->proxy->username . ':' . $this->proxy->password));
				$options['http']['header'] = $headers;
			}
		}		
		
		return $options;
	}	

	private function checkHttpCode($httpCode)
	{
		if ($httpCode == 403)
		{
			throw new GrabzItScrapeException('Potential DDOS Attack Detected. Please wait for your service to resume shortly. Also please slow the rate of requests you are sending to GrabzIt to ensure this does not happen in the future.', GrabzItException::NETWORK_DDOS_ATTACK);
		}
		else if ($httpCode >= 400)
		{
			throw new GrabzItScrapeException("A network error occured when connecting to the GrabzIt servers.", GrabzItException::NETWORK_GENERAL_ERROR);
		}
	}

	private function checkResponseHeader($header)
	{
		list($version,$httpCode,$msg) = explode(' ',$header[0], 3);
		$this->checkHttpCode($httpCode);
	}
}
?>